Fix for #17168 issue

askb opened the pull request:
Now `lastErr` saves the errors returned from subsequent attempts from `puller.Pull()`, when trying to look for a image in same endpoint with different version (v2,v1,v0). The returned error value is appended to the `lastErr`, which is returned to the caller and in turn to the client. This would ensure the client gets the correct error messages and not only the most recent (v0) message during the pull request.

fixes https://github.com/docker/docker/issues/17168

askb committed:
Fix for #17168 misleading pull error

This fix avoids overwritting the previous error messages, ensures the client gets the correct error messages and not just the most recent message during the pull request.
For this `var lastErr` replaced with a slice which acts as a temp place holder for the list of returned error messages for every attempt.
The slice is later joined and returned to the caller function after searching for the image with diffirent versions(v2,v1,v0).

Updated the code with check for no space left on device error occurance and prevent the
daemon on falling back to v1,v0.

Incorporated the comments from @calavera, @RichardScothern, @cpuguy83

Signed-off-by: Anil Belur <askb23@gmail.com>

calavera commented:
I think this could be easier to understand if you used an slice to append error strings. That way you wouldn't need to check whether it's the first error or not every time. Something like this:

```go
var errors []string // Initialize the errors
...
errors = append(errors, err.Error()) // When there is an error
...
if len(errors) > 0 { // At the end of the function
  return fmt.Error(strings.Join(errors, "\n"))
}
return nil
```

askb commented:
@calavera, thank you for input. I have incorporated suggested changes to use a slice, now the fix should reflect the changes.

askb commented:
Hi @RichardScothern: Now the fix is updated with a check to prevent fall back to V1,V0 on disk full error.

cpuguy83 commented on registry/registry.go:
```diff
@@ -219,6 +220,10 @@ func ContinueOnError(err error) bool {
 	case *client.UnexpectedHTTPResponseError:
 		return true
+	case error:
+		if val := strings.Contains(err.Error(), errNoSpaceOnDevice.Error()); val {
```
I'm not sure what the real error is that is returned here (maybe you can provide some insight on the actual error type here since you've been working on it?)... but we should compare this error to `syscall.ENOSPC` rather than defining `errNoSpaceOndevice`

askb replied:
@cpuguy83, The error returned is from `chrootarchive.invokeUnpack()` function, which returns a error string `"Untar re-exec error: exit status 1: output: write /lib/udev/hwdb.d/20-pci-vendor-model.hwdb: no space left on device"` instead of `syscall.ENOSPC`. Is there a better way to obtain the error `ENOSPC` in the caller function ?

cpuguy83 replied:
Ah no, this wouldn't be easy.
So we can do `strings.Contains(err.Error(), syscall.ENOSPC)`

cpuguy83 replied:
Ah no, this wouldn't be easy.
Instead we can do `strings.Contains(err.Error(), syscall.ENOSPC.Error())`

askb replied:
@cpuguy83 Thanks for the input, now the changes are updated.

calavera commented on registry/registry.go:
```diff
@@ -219,6 +220,10 @@ func ContinueOnError(err error) bool {
 	case *client.UnexpectedHTTPResponseError:
 		return true
+	case error:
+		if val := strings.Contains(err.Error(), strings.ToLower(syscall.ENOSPC.Error())); val {
```
you don't need to assign a value here, but I'm okay with it.

RichardScothern replied:
something like

```
patherr, ok := err.(*os.PathError)
if ok && patherr.Err == syscall.ENOSPC {
    return err
}
```

would be a bit cleaner than comparing strings.

cpuguy83 replied:
This is an error decoded from a reexec stdout, I odn't think you can compare the error here... but I could be wrong.

askb replied:
@RichardScothern @cpuguy83 is right `ENOSPC.Error()` string returned is a part of the larger string returned from `chrootarchive.invokeUnpack()` therefore we need to do sub-string search instead of comparing the error code, unfortunately. But going forward, it may better to return standardize error codes return values to caller. Thanks.

calavera commented:
LGTM. @RichardScothern what do you think about this change?

thaJeztah commented:
Thanks @askb!

I edited the top comment to include `fixes #17168`

RichardScothern commented:
Is it possible to add a test?

askb commented:
@RichardScothern: imu not sure if there is a way add to the existing cli integration test for this fix presently as no quota support is available for the containers, therefore mostly requires running manual tests. The test itself this requires limiting the disk space either on the container or vm hosting the container, except for keeping around 100 blocks system reserved space free, which needs to configured `tune2fs -r`. Please let me any others ways to add a test case ? I would be happy to give it a go. Thanks

vdemeester commented:
@askb we could add a test that is skipped if there is no quota suport on the test machine and thus be easy to test if there is support.. (using `testRequire` and co)

askb commented:
Hi @vdemeester `testRequire` is a good, wondering if there are some gaps in my understanding on how this can be implemented. Could we query a configuration file to determine if disk quotas for containers is available, something similar to what is done for `IPv6 testRequirement` ex: `cmd := exec.Command("test", "-f", "/proc/net/if_inet6")`. What flags or commands do we use to start or setup a container with disk quota, limiting the container to maybe `512MiB`, this would be required in the caller or test function (in integration-cli/docker_cli_{*}_test.go). Thanks.

cpuguy83 commented:
I'm not sure how feasible it is to add a test for running out of space...
Maybe we can create a small mount and point the daemon graph dir there?

askb commented:
@cpuguy83 I tried mounting a container into a loopback-fs, and faced some issues. I think this sounds good idea to point only the graph dir root into mount point, wihch avoids the need for doing a sync of the files from the container to the mount point. Will try it out and update here. Thanks

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

askb committed:
Adding test for the fix #17168.

The test case creates a mount point, starts the daemon with the
graph dir pointing to the mount and initates a pull request. We should
be able to check for the error message when the mount point gets filled
during pull.

Signed-off-by: Anil Belur <askb23@gmail.com>

askb commented:
@cpuguy83, Now adding test for the fix #17168. The test creates a mount point, starts the daemon with the graph dir pointing to the mount and initiates a pull request. We should be able to check for the error message when the mount point gets filled during pull request. Thanks

LK4D4 commented:
LGTM

thaJeztah commented:
Thanks for your contribution!

cpuguy83 commented:
:+1: Thanks for doing this. Sorry we've been silent during DockerCon!
Really good work on the test!
